packages/clerk-js/src/utils/captcha.ts

```diff
@@ -1,55 +1,132 @@
+   * Value must be a positive integer less than 900000.
+   * @default 8000
+   */
+  'retry-interval'?: number;
+  /**
+   * Automatically refreshes the token when it expires.
+   * Can take auto, manual or never.
+   * @default 'auto'
+   */
+  'refresh-expired'?: 'auto' | 'manual' | 'never';
+  /**
+   * A JavaScript callback invoked upon success of the challenge.
+   * The callback is passed a token that can be validated.
+   * @param token string
+   */
+  callback?: (token: string) => void;
+  /**
+   * A JavaScript callback invoked when there is an error (e.g. network error or the challenge failed)
+   * @param errorCode string
+   */
+  'error-callback'?: (errorCode: string) => void;
+}
+
+interface Turnstile {
+  execute: (container?: string | HTMLElement | null, params?: RenderOptions) => void;
+  render: (container?: string | HTMLElement | null, params?: RenderOptions) => string;
+  remove: (widgetId: string) => void;
+  reset: (widgetId: string) => void;
 }
 
 declare global {
   export interface Window {
-    turnstile: { execute: (container?: string | HTMLElement | null, params?: RenderOptions) => void };
+    turnstile: Turnstile;
   }
 }
 
 const WIDGET_CLASSNAME = 'clerk-captcha';
 
 export async function loadCaptcha(url: string) {
   if (!window.turnstile) {
-    await loadScript(url, { defer: true });
+    try {
+      await loadScript(url, { defer: true });
+    } catch (_) {
+      // Rethrow with specific message
+      clerkFailedToLoadThirdPartyScript('Cloudflare Turnstile');
+    }
   }
   return window.turnstile;
 }
+
 export const getCaptchaToken = async (captchaOptions: { siteKey: string; scriptUrl: string }) => {
   const { siteKey: sitekey, scriptUrl } = captchaOptions;
-  let captchaToken = '';
+  let captchaToken = '',
+    id = '';
 
   const div = document.createElement('div');
   div.classList.add(WIDGET_CLASSNAME);
   document.body.appendChild(div);
+
   const captcha = await loadCaptcha(scriptUrl);
+  let retries = 0;
 
-  const handleCaptchaTokenGeneration = (): Promise<string> => {
+  const handleCaptchaTokenGeneration = (): Promise<[string, string]> => {
     return new Promise((resolve, reject) => {
-      return captcha.execute(`.${WIDGET_CLASSNAME}`, {
-        sitekey,
-        retry: 'never',
-        callback: function (token: string) {
-          resolve(token);
-        },
-        'error-callback': function (err) {
-          reject(err);
-        },
-      });
+      try {
+        const id = captcha.render(`.${WIDGET_CLASSNAME}`, {
+          sitekey,
+          retry: 'never',
+          // 'retry-interval': 500,
+          'refresh-expired': 'auto',
+          callback: function (token: string) {
+            resolve([token, id]);
+          },
+          'error-callback': function (errorCode) {
+            /**
+             * By setting retry to 'never' the responsibility for implementing retrying is ours
```

Comment:
Why don't we switch to the default retry policy by CF?

Reply:
with retry:auto
once a challenge fails the  'error-callback' will be called.
and cf will decide if it want to retry.  For some error codes it will retry but for others it will not. (There is no way of knowing for which unless we transfer the logic that the mention in [their docs ](https://developers.cloudflare.com/turnstile/reference/client-side-errors/#error-codes)into our code which is not optimal).

It's important to note that if cf decides to retry, there is no stoping mechanism (e.g maxRetries) unless you explicitly remove the challenge

In our case their retry mechanism is not what we want as we simply need to attempt a few retries and then reject. With their implementation we cannot know when to reject because we don't know if the error-code  will trigger retries or not.
the default retry-interval is set to 8000ms (8s). I've set ours to 500ms which we can further decrease

:point_up: this is why i went with retry:never alongside manually calling .reset()
btw this is not a hack, [it is mentioned here](https://developers.cloudflare.com/turnstile/reference/client-side-errors/#retrying) in the "Retrying" seciont (edited)